File: Projects/usaco/prefix.java
/*
ID: rombin82
LANG: JAVA
TASK: prefix
*/

import java.io.*;
import java.util.*;
import java.util.stream.Collectors;

public class prefix {
	public static void main(String[] args) throws IOException {
		Scanner in = new Scanner(new File("prefix.in"));
		PrintWriter stdout = new PrintWriter(new File("prefix.out"));
		Runtime.getRuntime().addShutdownHook(new Thread(() -> stdout.flush()));
		String line;
		String sequence = "";
		List<String> primitives = new ArrayList<>();
		while (!(line = in.nextLine().trim()).equals("."))
			primitives.addAll(Arrays.asList(line.split(" ")));
		while (in.hasNext())
			sequence += in.next();
		int seqlen = sequence.length();
		int[] counts = new int[seqlen + 1];
		Set<Integer> lengths = primitives.stream().map(String::length).collect(Collectors.toSet());
		for (int i = seqlen - 1; i >= 0; i--)
			for (int l : lengths)
				if (l <= seqlen - i && primitives.contains(sequence.substring(i, i + l)))
					counts[i] = Math.max(counts[i], l + counts[i + l]);
		stdout.println(counts[0]);
	}
}

File: Projects/usaco/prefix.kt
/*
ID: rombin82
LANG: JAVA
TASK: prefix
*/

import java.io.*
import java.util.*
import java.util.stream.Collectors

object prefix {
    @Throws(IOException::class)
    fun main(args: Array<String>) {
        val stdin = Scanner(File("prefix.in"))
        val stdout = PrintWriter(File("prefix.out"))
        Runtime.getRuntime().addShutdownHook(Thread { stdout.flush() })
        var line: String
        var sequence = ""
        val primitives = ArrayList()
        while (!(line = stdin.nextLine().trim()).equals("."))
            primitives.addAll(Arrays.asList(line.split(" ")))
        while (stdin.hasNext())
            sequence += stdin.next()
        val seqlen = sequence.length()
        val counts = IntArray(seqlen + 1)
        val lengths = primitives.stream().map(???({ String.length() })).collect(Collectors.toSet())
        for (i in seqlen - 1 downTo 0)
            for (l in lengths)
                if (l <= seqlen - i && primitives.contains(sequence.substring(i, i + l)))
                    counts[i] = Math.max(counts[i], l + counts[i + l])
        stdout.println(counts[0])
    }
}
